OpenRA.Game/Traits/World/ActorMap.cs

```diff
@@ -402,13 +402,43 @@ public void Tick(Actor self)
 
 			addActorPosition.Clear();
 
+			UpdateTriggers();
+		}
+
+		public void UpdateTriggers()
+		{
 			foreach (var t in cellTriggers)
 				t.Value.Tick(this);
 
 			foreach (var t in proximityTriggers)
 				t.Value.Tick(this);
 		}
 
+		public void RemoveActor(Actor a)
+		{
+			foreach (var bin in bins)
+			{
+				var removed = bin.Actors.Remove(a);
+				if (removed > 0)
+					foreach (var t in bin.ProximityTriggers)
+						t.Dirty = true;
+			}
+
+			UpdateTriggers();
+		}
+
+		public void AddActor(Actor a)
+		{
+			var pos = a.CenterPosition;
+			var col = WorldCoordToBinIndex(pos.X).Clamp(0, cols - 1);
+			var row = WorldCoordToBinIndex(pos.Y).Clamp(0, rows - 1);
+			var bin = BinAt(row, col);
+
+			bin.Actors.Add(a);
+
+			UpdateTriggers();
```

Comment:
These still have the same problem of ticking the triggers at multiple and arbitrary times.